commit 4dee098023f73dd749ebc03eb87aa870e2b2c329
Author: Raffaello Giulietti <rgiulietti@openjdk.org>
Date:   Wed May 8 08:27:13 2024 +0000

    8330005: RandomGeneratorFactory.getDefault() throws exception when the runtime image only has java.base module
    
    Reviewed-by: jpai, alanb

diff --git a/src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java b/src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java
index 552d7b03d53..54d3b89173c 100644
--- a/src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java
+++ b/src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -139,8 +139,9 @@ public final class RandomGeneratorFactory<T extends RandomGenerator> {
          * @return Map of RandomGeneratorFactory classes.
          */
         private static Map<String, Provider<? extends RandomGenerator>> createFactoryMap() {
+            FactoryMapHolder.class.getModule().addUses(RandomGenerator.class);
             return ServiceLoader
-                .load(RandomGenerator.class)
+                .load(RandomGenerator.class, ClassLoader.getPlatformClassLoader())
                 .stream()
                 .filter(p -> !p.type().isInterface())
                 .collect(Collectors.toMap(p -> p.type().getSimpleName(), Function.identity()));
@@ -338,9 +339,6 @@ public final class RandomGeneratorFactory<T extends RandomGenerator> {
      * {@link RandomGenerator} that utilize the {@code name}
      * <a href="package-summary.html#algorithms">algorithm</a>.
      *
-     * @implSpec Availability is determined by RandomGeneratorFactory using the
-     * service provider API to locate implementations of the RandomGenerator interface.
-     *
      * @param name  Name of random number generator
      * <a href="package-summary.html#algorithms">algorithm</a>
      * @param <T> Sub-interface of {@link RandomGenerator} to produce
@@ -373,12 +371,9 @@ public final class RandomGeneratorFactory<T extends RandomGenerator> {
 
     /**
      * Returns a non-empty stream of available {@link RandomGeneratorFactory RandomGeneratorFactory(s)}.
-     *
+     * <p>
      * RandomGenerators that are marked as deprecated are not included in the result.
      *
-     * @implSpec Availability is determined by RandomGeneratorFactory using the service provider API
-     * to locate implementations of the RandomGenerator interface.
-     *
      * @return a non-empty stream of all available {@link RandomGeneratorFactory RandomGeneratorFactory(s)}.
      */
     public static Stream<RandomGeneratorFactory<RandomGenerator>> all() {
@@ -612,5 +607,3 @@ public final class RandomGeneratorFactory<T extends RandomGenerator> {
     }
 
 }
-
-
diff --git a/src/java.base/share/classes/java/util/random/package-info.java b/src/java.base/share/classes/java/util/random/package-info.java
index 2e56c1b6db0..3818bb55187 100644
--- a/src/java.base/share/classes/java/util/random/package-info.java
+++ b/src/java.base/share/classes/java/util/random/package-info.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -50,9 +50,7 @@
  * <p> The principal supporting class is {@link RandomGeneratorFactory}. This
  * can be used to generate multiple random number generators for a specific
  * algorithm. {@link RandomGeneratorFactory} also provides methods for
- * selecting random number generator algorithms. RandomGeneratorFactory
- * registers implementations of {@link RandomGenerator} interface using the
- * service provider API.
+ * selecting random number generator algorithms.
  *
  * <p> An important subsidiary interface is
  * {@link RandomGenerator.StreamableGenerator}, which provides methods for
@@ -631,4 +629,3 @@
  * @since   17
  */
 package java.util.random;
-
diff --git a/src/jdk.random/share/classes/jdk/random/L128X1024MixRandom.java b/src/java.base/share/classes/jdk/internal/random/L128X1024MixRandom.java
similarity index 99%
rename from src/jdk.random/share/classes/jdk/random/L128X1024MixRandom.java
rename to src/java.base/share/classes/jdk/internal/random/L128X1024MixRandom.java
index a612233f2ca..bc9d7065692 100644
--- a/src/jdk.random/share/classes/jdk/random/L128X1024MixRandom.java
+++ b/src/java.base/share/classes/jdk/internal/random/L128X1024MixRandom.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,7 +23,7 @@
  * questions.
  */
 
-package jdk.random;
+package jdk.internal.random;
 
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.random.RandomGenerator;
diff --git a/src/jdk.random/share/classes/jdk/random/L128X128MixRandom.java b/src/java.base/share/classes/jdk/internal/random/L128X128MixRandom.java
similarity index 99%
rename from src/jdk.random/share/classes/jdk/random/L128X128MixRandom.java
rename to src/java.base/share/classes/jdk/internal/random/L128X128MixRandom.java
index 1b7d25850a6..fa48b49ee0c 100644
--- a/src/jdk.random/share/classes/jdk/random/L128X128MixRandom.java
+++ b/src/java.base/share/classes/jdk/internal/random/L128X128MixRandom.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,7 +23,7 @@
  * questions.
  */
 
-package jdk.random;
+package jdk.internal.random;
 
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.random.RandomGenerator;
diff --git a/src/jdk.random/share/classes/jdk/random/L128X256MixRandom.java b/src/java.base/share/classes/jdk/internal/random/L128X256MixRandom.java
similarity index 99%
rename from src/jdk.random/share/classes/jdk/random/L128X256MixRandom.java
rename to src/java.base/share/classes/jdk/internal/random/L128X256MixRandom.java
index dad792d8c94..5b3c9978897 100644
--- a/src/jdk.random/share/classes/jdk/random/L128X256MixRandom.java
+++ b/src/java.base/share/classes/jdk/internal/random/L128X256MixRandom.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,7 +23,7 @@
  * questions.
  */
 
-package jdk.random;
+package jdk.internal.random;
 
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.random.RandomGenerator;
diff --git a/src/jdk.random/share/classes/jdk/random/L32X64MixRandom.java b/src/java.base/share/classes/jdk/internal/random/L32X64MixRandom.java
similarity index 98%
rename from src/jdk.random/share/classes/jdk/random/L32X64MixRandom.java
rename to src/java.base/share/classes/jdk/internal/random/L32X64MixRandom.java
index 1cb222e9cca..18997ed71df 100644
--- a/src/jdk.random/share/classes/jdk/random/L32X64MixRandom.java
+++ b/src/java.base/share/classes/jdk/internal/random/L32X64MixRandom.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,7 +23,7 @@
  * questions.
  */
 
-package jdk.random;
+package jdk.internal.random;
 
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.random.RandomGenerator;
diff --git a/src/jdk.random/share/classes/jdk/random/L64X1024MixRandom.java b/src/java.base/share/classes/jdk/internal/random/L64X1024MixRandom.java
similarity index 99%
rename from src/jdk.random/share/classes/jdk/random/L64X1024MixRandom.java
rename to src/java.base/share/classes/jdk/internal/random/L64X1024MixRandom.java
index 9ec81997bde..e4eb82d450e 100644
--- a/src/jdk.random/share/classes/jdk/random/L64X1024MixRandom.java
+++ b/src/java.base/share/classes/jdk/internal/random/L64X1024MixRandom.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,7 +23,7 @@
  * questions.
  */
 
-package jdk.random;
+package jdk.internal.random;
 
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.random.RandomGenerator;
diff --git a/src/jdk.random/share/classes/jdk/random/L64X128MixRandom.java b/src/java.base/share/classes/jdk/internal/random/L64X128MixRandom.java
similarity index 98%
rename from src/jdk.random/share/classes/jdk/random/L64X128MixRandom.java
rename to src/java.base/share/classes/jdk/internal/random/L64X128MixRandom.java
index 4188e85e89f..7dd757f312b 100644
--- a/src/jdk.random/share/classes/jdk/random/L64X128MixRandom.java
+++ b/src/java.base/share/classes/jdk/internal/random/L64X128MixRandom.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,7 +23,7 @@
  * questions.
  */
 
-package jdk.random;
+package jdk.internal.random;
 
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.random.RandomGenerator;
diff --git a/src/jdk.random/share/classes/jdk/random/L64X128StarStarRandom.java b/src/java.base/share/classes/jdk/internal/random/L64X128StarStarRandom.java
similarity index 98%
rename from src/jdk.random/share/classes/jdk/random/L64X128StarStarRandom.java
rename to src/java.base/share/classes/jdk/internal/random/L64X128StarStarRandom.java
index 139b27223b6..15b9bc312da 100644
--- a/src/jdk.random/share/classes/jdk/random/L64X128StarStarRandom.java
+++ b/src/java.base/share/classes/jdk/internal/random/L64X128StarStarRandom.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,7 +23,7 @@
  * questions.
  */
 
-package jdk.random;
+package jdk.internal.random;
 
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.random.RandomGenerator;
diff --git a/src/jdk.random/share/classes/jdk/random/L64X256MixRandom.java b/src/java.base/share/classes/jdk/internal/random/L64X256MixRandom.java
similarity index 99%
rename from src/jdk.random/share/classes/jdk/random/L64X256MixRandom.java
rename to src/java.base/share/classes/jdk/internal/random/L64X256MixRandom.java
index 0c0cbb43687..bb605df5fdc 100644
--- a/src/jdk.random/share/classes/jdk/random/L64X256MixRandom.java
+++ b/src/java.base/share/classes/jdk/internal/random/L64X256MixRandom.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,7 +23,7 @@
  * questions.
  */
 
-package jdk.random;
+package jdk.internal.random;
 
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.random.RandomGenerator;
diff --git a/src/jdk.random/share/classes/jdk/random/Xoroshiro128PlusPlus.java b/src/java.base/share/classes/jdk/internal/random/Xoroshiro128PlusPlus.java
similarity index 99%
rename from src/jdk.random/share/classes/jdk/random/Xoroshiro128PlusPlus.java
rename to src/java.base/share/classes/jdk/internal/random/Xoroshiro128PlusPlus.java
index ba8daaf7246..d20b3cf9e6a 100644
--- a/src/jdk.random/share/classes/jdk/random/Xoroshiro128PlusPlus.java
+++ b/src/java.base/share/classes/jdk/internal/random/Xoroshiro128PlusPlus.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,7 +23,7 @@
  * questions.
  */
 
-package jdk.random;
+package jdk.internal.random;
 
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.random.RandomGenerator;
diff --git a/src/jdk.random/share/classes/jdk/random/Xoshiro256PlusPlus.java b/src/java.base/share/classes/jdk/internal/random/Xoshiro256PlusPlus.java
similarity index 99%
rename from src/jdk.random/share/classes/jdk/random/Xoshiro256PlusPlus.java
rename to src/java.base/share/classes/jdk/internal/random/Xoshiro256PlusPlus.java
index 5c9cc606597..fdefe02dd03 100644
--- a/src/jdk.random/share/classes/jdk/random/Xoshiro256PlusPlus.java
+++ b/src/java.base/share/classes/jdk/internal/random/Xoshiro256PlusPlus.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -23,7 +23,7 @@
  * questions.
  */
 
-package jdk.random;
+package jdk.internal.random;
 
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.random.RandomGenerator;
diff --git a/src/java.base/share/classes/module-info.java b/src/java.base/share/classes/module-info.java
index 9d4a794de1a..06bea2dfd10 100644
--- a/src/java.base/share/classes/module-info.java
+++ b/src/java.base/share/classes/module-info.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2014, 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2014, 2024, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -253,8 +253,16 @@ module java.base {
         jdk.jfr;
     exports jdk.internal.util.xml.impl to
         jdk.jfr;
-    exports jdk.internal.util.random to
-        jdk.random;
+    exports jdk.internal.util to
+        java.desktop,
+        java.prefs,
+        java.security.jgss,
+        java.smartcardio,
+        jdk.charsets,
+        jdk.internal.vm.ci,
+        jdk.jlink,
+        jdk.jpackage,
+        jdk.net;
     exports sun.net to
         java.net.http,
         jdk.naming.dns;
@@ -383,7 +391,6 @@ module java.base {
     uses java.time.chrono.AbstractChronology;
     uses java.time.chrono.Chronology;
     uses java.time.zone.ZoneRulesProvider;
-    uses java.util.random.RandomGenerator;
     uses java.util.spi.CalendarDataProvider;
     uses java.util.spi.CalendarNameProvider;
     uses java.util.spi.CurrencyNameProvider;
@@ -411,6 +418,16 @@ module java.base {
     provides java.util.random.RandomGenerator with
         java.security.SecureRandom,
         java.util.Random,
-        java.util.SplittableRandom;
+        java.util.SplittableRandom,
+        jdk.internal.random.L32X64MixRandom,
+        jdk.internal.random.L64X128MixRandom,
+        jdk.internal.random.L64X128StarStarRandom,
+        jdk.internal.random.L64X256MixRandom,
+        jdk.internal.random.L64X1024MixRandom,
+        jdk.internal.random.L128X128MixRandom,
+        jdk.internal.random.L128X256MixRandom,
+        jdk.internal.random.L128X1024MixRandom,
+        jdk.internal.random.Xoroshiro128PlusPlus,
+        jdk.internal.random.Xoshiro256PlusPlus;
 
 }
diff --git a/src/jdk.random/share/classes/module-info.java b/src/jdk.random/share/classes/module-info.java
deleted file mode 100644
index 1b2193ef557..00000000000
--- a/src/jdk.random/share/classes/module-info.java
+++ /dev/null
@@ -1,64 +0,0 @@
-/*
- * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
- * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
- *
- * This code is free software; you can redistribute it and/or modify it
- * under the terms of the GNU General Public License version 2 only, as
- * published by the Free Software Foundation.  Oracle designates this
- * particular file as subject to the "Classpath" exception as provided
- * by Oracle in the LICENSE file that accompanied this code.
- *
- * This code is distributed in the hope that it will be useful, but WITHOUT
- * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
- * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
- * version 2 for more details (a copy is included in the LICENSE file that
- * accompanied this code).
- *
- * You should have received a copy of the GNU General Public License version
- * 2 along with this work; if not, write to the Free Software Foundation,
- * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
- *
- * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
- * or visit www.oracle.com if you need additional information or have any
- * questions.
- */
-
-import jdk.internal.util.random.RandomSupport;
-
-/**
- * Defines implementations of the
- * {@linkplain java.util.random.RandomGenerator RandomGenerator Interface}.
- *
- * @provides jdk.random.L128X1024MixRandom
- * @provides jdk.random.L128X128MixRandom
- * @provides jdk.random.L128X256MixRandom
- * @provides jdk.random.L32X64MixRandom
- * @provides jdk.random.L64X1024MixRandom
- * @provides jdk.random.L64X128MixRandom
- * @provides jdk.random.L64X128StarStarRandom
- * @provides jdk.random.L64X256MixRandom
- * @provides jdk.random.Xoroshiro128PlusPlus
- * @provides jdk.random.Xoshiro256PlusPlus
- *
- * @use java.util.random.RandomGenerator
- * @use jdk.internal.util.random.RandomSupport
- *
- * @moduleGraph
- * @since 16
- */
-module jdk.random {
-    exports jdk.random to
-            java.base;
-
-    provides java.util.random.RandomGenerator with
-        jdk.random.L32X64MixRandom,
-        jdk.random.L64X128MixRandom,
-        jdk.random.L64X128StarStarRandom,
-        jdk.random.L64X256MixRandom,
-        jdk.random.L64X1024MixRandom,
-        jdk.random.L128X128MixRandom,
-        jdk.random.L128X256MixRandom,
-        jdk.random.L128X1024MixRandom,
-        jdk.random.Xoroshiro128PlusPlus,
-        jdk.random.Xoshiro256PlusPlus;
-}
